nucypher/utilities/clouddeploy.py

```diff
@@ -0,0 +1,798 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+import copy
+import os
+import re
+import json
+import maya
+import time
+from base64 import b64encode
+from jinja2 import Template
+import requests
+
+from ansible.playbook import Playbook
+from ansible.parsing.dataloader import DataLoader
+from ansible.executor.task_queue_manager import TaskQueueManager
+from ansible.inventory.manager import InventoryManager
+from ansible.plugins.callback import CallbackBase
+from ansible.vars.manager import VariableManager
+from ansible.playbook.play import Play
+from ansible.executor.playbook_executor import PlaybookExecutor
+from ansible import context as ansible_context
+from ansible.module_utils.common.collections import ImmutableDict
+
+from nucypher.config.constants import DEFAULT_CONFIG_ROOT
+from nucypher.blockchain.eth.clients import PUBLIC_CHAINS
+from nucypher.blockchain.eth.networks import NetworksInventory
+
+NODE_CONFIG_STORAGE_KEY = 'worker-configs'
+URSULA_PORT = 9151
+PROMETHEUS_PORTS = [9101]
+
+
+ansible_context.CLIARGS = ImmutableDict(
+    {
+        'syntax': False,
+        'start_at_task': None,
+        'verbosity': 0,
+        'become_method': 'sudo'
+    }
+)
+
+class AnsiblePlayBookResultsCollector(CallbackBase):
+    """
+    执行playbook的回调类
+    """
```

Comment:
What's going on here?

Reply:
ha. For some reason all of the best examples of writing results collectors for Ansible are from Chinese blogs... and I guess I must have done a little copy/paste